### DISCUSSION TITLE
syntax for tracking combining of ingredients

### DISCUSSION THREAD

**Discussion by aspiers** [cooklang/spec#62] [Ideas]
Pretty much every recipe out there involves combining two or more ingredients in some way.  I'm not counting "fry an egg" as a recipe ;-)

It's [long been my dream](https://twitter.com/adamspiers/status/1463473628197138434) to have a [d3](https://d3js.org/)-based tool which automatically rendered an "ingredients combination diagram" like the one below from a (mostly human-readable) recipe, clearly showing the critical path, and could even dynamically adjust the rendering if things didn't go to the original plan.  Here's my Xmas dinner sketch from 2015:

![Christmas Dinner](https://user-images.githubusercontent.com/100738/143334793-174dede5-4203-4b74-bb4e-9db123022ff9.png)

So you can imagine how delighted I was to discover cooklang earlier today.

But in order to realise this dream, cooklang would need to somehow track in its syntax the point at which A is combined with B, where A and B can both be either a single ingredient, or a previously combined set of ingredients.  When A is combined with B, what data might need tracking?  I can think of the following:

1. A container / holder, e.g. a salad bowl / cup / saucepan / baking tray etc.
2. The verb describing the act of combining, e.g. "mix", or "stir", or "add"
3. Some extra details regarding *how* to do it, e.g. if the verb is "stir", then the details might be "for 30 seconds"

I don't have any concrete suggestions yet, but I'm hoping this will provoke some useful thoughts and discussions yet.  My instinct is that cooklang should have syntax to "mark up" the container and verb, but the extra details could be left as unparsed natural language, unless perhaps they include timings.

**Comment by aspiers** [comment]
Oops, perhaps I should have filed a Discussion topic for this rather than an issue.  Happy to move if that makes sense.

**Comment by chrbala** [comment]
I'm interested in this as well. I would like this to be decoupled from cooklang or any other recipe format so that it can be used with other similar systems.

A couple things:
* How should something like this work on mobile? It will be far more useful if it's available on small screens, so it should be possible to navigate by small sections of the recipe.
* Formally, I'd say that recipes form a directed acyclic graph (DAG). Many recipes form tree structures and some are simple lines. The most complex recipes form a DAG though, since items can be dependencies on multiple other items. It looks like the honey/coriander portion of the diagram above fits the DAG description. It can be simpler to group clusters of DAG structures into larger nodes such that as much of the recipe can be represented as a tree as possible. DAG UIs are difficult, so boxing the complexity in to small components will reduce the overall complexity of the whole recipe.
* Time is an important factor. To order steps in a DAG correctly, some concept of time may be needed. For example, if a sauce must be made before the pasta, the sauce takes 5 minutes, and the pasta takes 30, the sauce can be made while the pasta cooks. The traits I've identified for time management are (1) time taken (2) how fragile asynchronous tasks are, like frying an egg vs boiling water, and (3) how much overhead asynchronous tasks take, like regularly stirring sauce on the stove so it doesn't burn vs baking bread without much oversight.
* Being prescriptive in _telling_ cooks how they should cook is likely not the right move. Part of the role of the system is to surface critical information at the right time so that cooks can decide what is best based on their specific situation. 
* I'm not sure cooklang could represent all of these pieces, since it's a human-readable format. If we want to support these features in a UI, it may make sense to decouple from cooklang into cooklang (backend) and the UI (frontend) so that the frontend can support more complex modeling from other applications and cooklang can send a simple representation. I'm happy to be proven wrong here, it just seems like there is _a lot_ to handle this properly.

The nearest thing I've seen to a horizontally scalable UI like this is [Partition](http://mbostock.github.io/d3/talk/20111018/partition.html), which is built from D3. Recipes like those found on [cooking for engineers](http://www.cookingforengineers.com/recipe/158/Dark-Chocolate-Brownies) could be represented in that way on both small and large screens, though it probably wouldn't work very well for DAG shaped data.

<img width="633" alt="Screen Shot 2021-11-25 at 12 17 23 AM" src="https://user-images.githubusercontent.com/6834804/143404586-47a9988c-968f-4945-9366-dc0dc430a03d.png">

I have experimented a bit with DAG shaped data in UIs, as shown in the below videos. It's certainly less complex than the use cases I mention above. I made this a long time ago and found that it wasn't a sufficient model for representing what I needed.

This one shows a hierarchy and two runs with the steps completed in a different order.
https://user-images.githubusercontent.com/6834804/143406458-229b7716-bae8-4638-9cad-ad01fd933027.mov

This one does the same as above, but flattens the list.
https://user-images.githubusercontent.com/6834804/143406622-d3de1c7a-f61b-4257-aeb7-e1f4247d6146.mov

This one flattens the list, but leaves the instructions in place after completion.
https://user-images.githubusercontent.com/6834804/143406588-1f04a649-121d-4de6-b9e5-d21b18bda382.mov

I have a live version of these example videos [available here](https://banker-porcupine-34023.netlify.app/?knob-showBlocked=true&knob-showDone=true&knob-flatten=true&selectedKind=Recipe&selectedStory=Recipe&full=1&down=1&left=1&panelRight=0&downPanel=kadirahq%2Fstorybook-addon-knobs). I was not diligent with CI or preservation though, so I don't have a way to make changes, and might not even have the source code anymore.

**Comment by aspiers** [comment]
Excited we're on the same page here!

[**@chrbala**](https://github.com/chrbala) commented on [November 25, 2021 8:36 AM](https://github.com/cooklang/spec/issues/36#issuecomment-978953310 "2021-11-25T08:36:30Z - Replied by Github Reply Comments"):
> I'm interested in this as well. I would like this to be decoupled from cooklang or any other recipe format so that it can be used with other similar systems.
> 
> A couple things:
> 
> *   How should something like this work on mobile? It will be far more useful if it's available on small screens, so it should be possible to navigate by small sections of the recipe.

I expect it should be pretty easy to get d3 to allow panning and zooming via touch and multi-touch, e.g.

https://github.com/doter1995/d3-zoom-xy

> *   Formally, I'd say that recipes form a directed acyclic graph (DAG). Many recipes form tree structures and some are simple lines. The most complex recipes form a DAG though, since items can be dependencies on multiple other items.
>
> It looks like the honey/coriander portion of the diagram above fits the DAG description.

Yes, exactly.  Another case might be cooking something in the juices of something else, then removing it from the pan, doing something else in the pan, then recombining everything later.  If we exclude those kinds of cases then I suspect a tree structure is sufficient.  But of course we want cooklang to be able to cope with all cases, so I agree that it's a DAG.

> It can be simpler to group clusters of DAG structures into larger nodes such that as much of the recipe can be represented as a tree as possible. DAG UIs are difficult, so boxing the complexity in to small components will reduce the overall complexity of the whole recipe.

Some kind of high-level grouping (e.g. meat / vegetables / sauce, even dessert being prepared in parallel) would be really helpful.

> *   Time is an important factor. To order steps in a DAG correctly, some concept of time may be needed. For example, if a sauce must be made before the pasta, the sauce takes 5 minutes, and the pasta takes 30, the sauce can be made while the pasta cooks. The traits I've identified for time management are (1) time taken (2) how fragile asynchronous tasks are, like frying an egg vs boiling water, and (3) how much overhead asynchronous tasks take, like regularly stirring sauce on the stove so it doesn't burn vs baking bread without much oversight.

Exactly!!  That was my main motivation for this whole thing actually, because I'm a terrible cook, and timing is the main area where things can go wrong...

> *   Being prescriptive in _telling_ cooks how they should cook is likely not the right move. Part of the role of the system is to surface critical information at the right time so that cooks can decide what is best based on their specific situation.

Violently agree :-)

> *   I'm not sure cooklang could represent all of these pieces, since it's a human-readable format. If we want to support these features in a UI, it may make sense to decouple from cooklang into cooklang (backend) and the UI (frontend) so that the frontend can support more complex modeling from other applications and cooklang can send a simple representation. I'm happy to be proven wrong here, it just seems like there is _a lot_ to handle this properly.

Yes, it could be that it makes more sense to model the DAG outside the main recipe, in the same way that cooklang already has shopping lists separately.  But I have a hunch that it *might* be possible to do inline without harming the readability too much.

> The nearest thing I've seen to a horizontally scalable UI like this is [Partition](http://mbostock.github.io/d3/talk/20111018/partition.html), which is built from D3.

I was thinking more along the lines of https://observablehq.com/@d3/collapsible-tree or https://observablehq.com/@d3/tree (with left and right flipped, of course), except making it a force-directed graph rather than a static tree.  In fact I built one of these with d3 years ago: https://github.com/aspiers/git-deps

Aesthetically, I would love something with bright colours and connecting lines which aren't super thin e.g. try to imagine a cross between one of the two trees and https://observablehq.com/@d3/sankey

My original idea was to use the aesthetics of [the London Underground tube map](https://duckduckgo.com/?q=tube+map+london&t=brave&iax=images&ia=images), which has separate coloured lines joining in places.

> Recipes like those found on [cooking for engineers](http://www.cookingforengineers.com/recipe/158/Dark-Chocolate-Brownies) could be represented in that way on both small and large screens, though it probably wouldn't work very well for DAG shaped data.
> 
> [![Screen Shot 2021-11-25 at 12 17 23 AM](https://user-images.githubusercontent.com/6834804/143404586-47a9988c-968f-4945-9366-dc0dc430a03d.png)](https://user-images.githubusercontent.com/6834804/143404586-47a9988c-968f-4945-9366-dc0dc430a03d.png)

Yeah, my friend who has an engineering mind and is an amazing cook introduced me to that really cool site a few years back, when I shared this whole idea with him :-)
 
> I have experimented a bit with DAG shaped data in UIs, as shown in the below videos. It's certainly less complex than the use cases I mention above. I made this a long time ago and found that it wasn't a sufficient model for representing what I needed.
> 
> This one shows a hierarchy and two runs with the steps completed in a different order.  
> [user-images.githubusercontent.com/6834804/143406458-229b7716-bae8-4638-9cad-ad01fd933027.mov](https://user-images.githubusercontent.com/6834804/143406458-229b7716-bae8-4638-9cad-ad01fd933027.mov)
> 
> This one does the same as above, but flattens the list.  
> [user-images.githubusercontent.com/6834804/143406622-d3de1c7a-f61b-4257-aeb7-e1f4247d6146.mov](https://user-images.githubusercontent.com/6834804/143406622-d3de1c7a-f61b-4257-aeb7-e1f4247d6146.mov)
> 
> This one flattens the list, but leaves the instructions in place after completion.  
> [user-images.githubusercontent.com/6834804/143406588-1f04a649-121d-4de6-b9e5-d21b18bda382.mov](https://user-images.githubusercontent.com/6834804/143406588-1f04a649-121d-4de6-b9e5-d21b18bda382.mov)
> 
> I have a live version of these example videos [available here](https://banker-porcupine-34023.netlify.app/?knob-showBlocked=true&knob-showDone=true&knob-flatten=true&selectedKind=Recipe&selectedStory=Recipe&full=1&down=1&left=1&panelRight=0&downPanel=kadirahq%2Fstorybook-addon-knobs). I was not diligent with CI or preservation though, so I don't have a way to make changes, and might not even have the source code anymore.

Ah, nice!

**Comment by chrbala** [comment]
> I expect it should be pretty easy to get d3 to allow panning and zooming via touch and multi-touch  

I think the risk here for usage on mobile is that navigation could be difficult. If the primary way of navigation is panning and zooming, it may be easy to get lost on smaller screens, even if it's technically possible to move around. What I like about the Partition example is that it can show a fixed number of levels at once. A graph-based system could do something similar, focusing 1 node and showing adjacent nodes which bring focus when clicked or tapped.

> Yes, it could be that it makes more sense to model the DAG outside the main recipe, in the same way that cooklang   already has shopping lists separately. But I have a hunch that it might be possible to do inline without harming the readability too much.  

Perhaps the cooklang DSL part here could be something like adding tags to steps in the recipe so arbitrary metadata can live elsewhere. Something like `[boil] Boil @eggs{2} for ~eggs{3%minutes}.` where `boil` could be used for tagging and processing externally. They could either be manually authored or automatically generated. This could be used regardless of whether the metadata is stored in the recipe itself or not.

**Comment by chrbala** [comment]
The other question I'd posit here is: what is the purpose of this system? It could be for:
* live information while cooking
* helping understand recipes before cooking
* helping edit DAG shaped recipes
* helping decide the order of cooking before starting

This could be used for any/all of the above, and would look different depending on which it's trying to support.

**Comment by dubadub** [comment]
This is a great thread to read: many interesting ideas and conversation goes really deep. 

While I think complete computational representation of a recipe may not be as practical for the most of the Cooklang users, I often return to DAG nature of recipes in the context of this problem https://github.com/cooklang/spec/issues/10. Every recipe contains multiple steps, each step is sort of a recipe too (not finished) and often ingredients are recipes too (like bolognese sauce).

I'm kind of stuck and don't know how to put things together in UI, similarly as you mentioned.

**Comment by aspiers** [comment]
Great thread indeed! :+1: 

[**@chrbala**](https://github.com/chrbala) commented on [November 26, 2021 3:42 AM](https://github.com/cooklang/spec/issues/36#issuecomment-979672201 "2021-11-26T03:42:00Z - Replied by Github Reply Comments"):
> The other question I'd posit here is: what is the purpose of this system? It could be for:
> 
> *   live information while cooking
> *   helping understand recipes before cooking
> *   helping edit DAG shaped recipes
> *   helping decide the order of cooking before starting
> 
> This could be used for any/all of the above, and would look different depending on which it's trying to support.

Totally agree!

I don't think there should be any problem with panning and zooming on smaller devices, there are several simple techniques which can make this an extremely pleasant UX.

I am also more optimistic regarding encapsulating the DAG within the language syntax.  I think there should be a simple way to indicate the merging (or splitting) of two (or possibly more) ingredients, and that is all which would be required.  The user doesn't need to think of it as a DAG at all.

Here are some ideas off the top of my head.  They probably need a lot of work before being usable though!

For example, taking the screenshot from the home page:

    Then add @salt and @ground black pepper{} to taste

could become something like

    Then add[@salt and @ground black pepper{}] to taste

The following could also be accepted:

    Then add[@salt, and @ground black pepper{}] to taste
    Then add[@salt, herbs, and @ground black pepper{}] to taste

The parser would automatically know that these ingredients should be added to whichever ingredient or container was mentioned last, or the destination could be referenced explicitly.

Some extra minimal markup would be required for tracking those, as well as the action (verb) and/or container involved during the combination.  Maybe we could use `+` for combination verbs, `>` to represent the destination (a container, e.g. pan / pot / blender / cup / jug), and `=` for verbs which represent actions on an existing combination, e.g.

```
-- Source: https://www.jamieoliver.com/recipes/eggs-recipes/easy-pancakes/

+Crack the @eggs{3} into a >#blender, then +add [the @flour{125%g}, @milk{250%ml} and @sea salt{1%pinch}], and =[blitz until smooth].

+Pour into a >#bowl and =[leave to stand for ~{15%minutes}].

=Melt the @butter (or a drizzle of @oil if you want to be a bit healthier) in a >#large non-stick frying pan{} on a medium heat, then =tilt the pan so the butter coats the surface.

=Pour in 1 ladle of batter and =tilt again, so that the batter spreads all over the base, then cook for 1 to 2 minutes, or until it starts to come away from the sides.

Once golden underneath, =flip the pancake over and cook for 1 further minute, or until cooked through.

Serve straightaway with your favourite topping. -- Add your favorite topping here to make sure it's included in your meal plan!
```

This is far from complete and it's not 100% clear how a machine could infer the DAG in all cases, but maybe something along these lines could work.

Maybe verbs such as crack / pour / melt could even be autodetected.

**Comment by tobei** [comment]
This matches something i proposed in the idea/discussion part and exaxctly the way i draw my recipe on paper before i cook them... i find it so much easier than a step by step approach telling me "mix all the dry ingredients" then "add milk" then "wait for 30 min frigge to rise" then "blend with whip until solid consistency" to have my ingredients on their own line combining with each other graphically with timing, instructions, tools and verbs used on those links, forming compound that are then later on assembled to make bigger compound.

One of the big advantage of that graph/tree based approach is refactoring : it becomes easy to make a whole branch to split into its own recipe, reference an external recipe as a subgraph, or merge recipe together for sharing for example. As long as each subtree/subrecipe contributes to the global ingredients/grocery list.

I always think recipe as a tree and almost never as step 1, 2, 3, because it's never evident for me by reading alone what steps are parallel or when they reuse all or part of a previous step result or where a particular ingredients gets used or easily forgotten...

**Comment by aspiers** [comment]
Check out this great thread! https://twitter.com/hydrosquall/status/1486157532968148996 - seems there is prior art in this space, and more like-minded folks :-)

**Reply by tobei** [reply]
Wow, that's exactly what i have been doing all along to translate a written recipe into something I can actually follow along in my kitchen ! I'm totally unable to follow a step by step text recipe without drawing that diagram first. Two interresting things from that thread:
- It is possible from a tree based recipe to automatically rewrite it into a step by step linear recipe, but not the other way around.
- I didn't think about items that can be separated, like egg becoming yolk and white used in two different preparation (split vs combine)
- I love that steps like wait time are also part of the design ;-)
